### PR TITLE
feat(pkg/trie): TrieDB commits

### LINKED ISSUE #3927: (pkg/trie): Implement node encoding with new node enums
Since we change the node structure for trieDB we need to implement the new nodes encoding in order to commit the trie data in the database.

### LINKED ISSUE #3993: (pkg/trie): Implement DB commit
We have to implement the `commit` method in `TrieDB` to flush the temporal changes into our DB.
This method will be called in every `Hash` call so every time we need to calculate the trie root hash we are going to commit the pending changes

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

This PR adds commit capability for `trieDB` to flush the temporary changes into the underlying DB. 
It works by traversing the temporary changes contained in `deathRow` for deletions and in `storage` for upserts, to delete, insert, or update them in our DB.

### Summary of changes

- [x] Add new nodes encoding
- [x] Commit changes in our DB
    - [x]  Deletes
    - [x]  Upserts
- [x] Tests
    - [x] Check nodes / values with V0 are working ok
    - [x] Check nodes / values using V1 are working ok

## Tests

```sh
make test
```

## Issues

closes: #3993, #3927

### CODE REVIEW

File: pkg/trie/triedb/node.go
Comment (jimjbrettj) [inline pkg/trie/triedb/node.go]: Will this be addressed in this PR?

File: pkg/trie/triedb/node.go
Comment (jimjbrettj) [inline pkg/trie/triedb/node.go:285]: Im pretty sure eclesio already has asked about this but I wanted to double check, this is an okay situation where the value is nil but there are children?

File: pkg/trie/triedb/node.go
Comment (jimjbrettj) [inline pkg/trie/triedb/node.go]: Is this still an active todo?

File: pkg/trie/triedb/node.go
Comment (jimjbrettj) [inline pkg/trie/triedb/node.go]: nit/what is oindex? maybe a more clear name could be used

File: pkg/trie/triedb/node.go
Comment (jimjbrettj) [inline pkg/trie/triedb/node.go]: why oslice?

Comment (jimjbrettj) [inline pkg/trie/triedb/node.go]: I think would be useful to add a comment here describing what function types purpose is

File: pkg/trie/db/db.go
Comment (jimjbrettj) [inline pkg/trie/db/db.go:25]: Since Database if the getter plus the putter, and the writer used hear already contains the put method, and we just remove this interface and change `Database` to be `type RWDatabase interface {
	DBGetter
	Writer
}` and just use that everywhere instead?

File: pkg/trie/triedb/triedb.go
Comment (jimjbrettj) [inline pkg/trie/triedb/triedb.go]: Should we use our internal logger here instead?

File: pkg/trie/triedb/triedb_test.go
Comment (jimjbrettj) [inline pkg/trie/triedb/triedb_test.go:711]: Maybe have a test for getting a value that was committed and then deleted? Making sure that we get the correct data when a node doesn't exist anymore

File: pkg/trie/triedb/util_test.go
Comment (jimjbrettj) [inline pkg/trie/triedb/util_test.go:66]: Does this need to be implemented? Should we panic in the meantime?

**Review by jimjbrettj — COMMENTED** [review]
Nice work! Just a few comments and suggestions

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Thanks for mentioning it! I'll address it in a different PR with a refactor

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go:285]: Yeah sure that's possible

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Removed! I already test it

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: All renamed now

Comment (dimartiro) [reply pkg/trie/triedb/node.go]: I think it's more clear now that I renamed it, but let me know what you think

File: pkg/trie/db/db.go
Comment (dimartiro) [reply pkg/trie/db/db.go:25]: The difference is that `RWDatabase` also includes the `Del`

File: pkg/trie/triedb/triedb.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb.go]: I took a similar approach than @timwu20 with https://github.com/ChainSafe/gossamer/pull/4015

File: pkg/trie/triedb/util_test.go
Comment (dimartiro) [reply pkg/trie/triedb/util_test.go:66]: Since it is an in-memory db we don't have things to flush 😄

File: pkg/trie/triedb/triedb_test.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb_test.go:711]: Added! and thanks that I found a bug

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Renamed

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go:45]: this function has 2 switches, but I think you can merge both.

Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go:45]: question: the `newValueRef` can be a value that a node in the trie holds? or is this a ref only to hash of child nodes?

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]: this method can be removed since `getNodeData` exists

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]: From what I understood, theses shifts can be replaced by a cohesive
```suggestion
const emptyTrieBytes = []byte{0}
```

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]: could you define what a childFunc does?

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]: why are you taking the last item of the partial key and are using it in the `childF` function?

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]:
```suggestion
	} else if value.IsHashed() {
	   ...
	} else {
	   ...
	}
```

File: pkg/trie/triedb/triedb.go
Comment (EclesioMeloJunior) [reply pkg/trie/triedb/triedb.go]: I would suggest using `logger.Debug("Committing trie changes to db")` so we can have it logged following our logging style

File: pkg/trie/triedb/triedb.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/triedb.go]: I would suggest using `logger.Debug`

File: pkg/trie/triedb/triedb.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/triedb.go]: could you open a batch, do the modifications and then flush applying all these modifications to the database at once? I would say it is better than hit the db every time (you might want to rollback if something went wrong)

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: pkg/trie/triedb/triedb.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb.go]: Changed to use pkg level logger! thks

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go:45]: Done! I simplified that method

Comment (dimartiro) [reply pkg/trie/triedb/node.go:45]: It is a V1 value, so it's the hash pointing to the real value in our db. In other words, is a reference (hash) to the key that holds that value in our DB

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Removed, thks!

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Actually if we want to replace it should be by:
```go
const emptyTrieBytes = byte(0)
```

Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Replaced 😄

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Sure, this is just an alias for a closure we execute when we are storing child nodes, you can see an example [here](https://github.com/ChainSafe/gossamer/blob/0eebe2d8cd2dc4c2befcca60285c4d6b297d5b26/pkg/trie/triedb/triedb.go#L807-L841)

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: This is to calculate the partial key of the child nodes

File: pkg/trie/triedb/triedb.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb.go]: Sure! good idea

File: pkg/trie/triedb/node.go
Comment (P1sar) [inline pkg/trie/triedb/node.go]: I would rather recommend creating an issue for this right away, there is nothing more permanent as something expected to be temporal

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [reply pkg/trie/triedb/node.go]: But is there any need for the slice notation `[:]`? can you just get the last element `arr[len(arr)-1]`?

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [reply pkg/trie/triedb/node.go]: I see, can we rename it to `onChildStoreFn`, since this function stores functions, also do it need to be a closure, cannot it be an actual fn or routine that is called? I think that the need for a closure is when there are differences on how you implement something, eg. a predicate closure when might exists different conditions?

Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Renamed! we use a closure to use free variables, for example [here](https://github.com/ChainSafe/gossamer/blob/0eebe2d8cd2dc4c2befcca60285c4d6b297d5b26/pkg/trie/triedb/triedb.go#L808) with the `prefixKey`.
I can create an issue to revisit it later if you want

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: I've been checking and don't make sense to move this code, so I'll remove the TODO

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Oh! I got your point, I think we have a bug here and didn't arise because we are using small partialkeys in our tests, I'll take a look

Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Fixed!

File: pkg/trie/triedb/triedb.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb.go]: I added support for batches, not only for deletes but also for the inserts / updates. Thanks for the reminder.

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]:
```suggestion
			if v.hash != common.EmptyHash && v.hash != cr.hash {
					panic("hash mismatch")
			}
```

File: pkg/trie/triedb/node.go
Comment (EclesioMeloJunior) [inline pkg/trie/triedb/node.go]: in the prev check you make sure that `v.hash` and `cr.hash` are the same, so is this assign needed?

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM, just small nits

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: There was an else missing, [this is the logic](https://github.com/paritytech/trie/blob/c3348e4b335443c17586a8b7491ba893cfde76a6/trie-db/src/triedbmut.rs#L166-L170) in trie crate.
I added it and this will also solves your [second](https://github.com/ChainSafe/gossamer/pull/4039/files#r1653624350) comment

File: pkg/trie/triedb/node.go
Comment (dimartiro) [reply pkg/trie/triedb/node.go]: Solved with this [commit](https://github.com/ChainSafe/gossamer/pull/4039/commits/46bfce7a153db5a9c46cae9ab4021addcd6f2562)

File: pkg/trie/triedb/codec/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/codec/node.go:77]: Why panic and not an error?

Unless it is an error that would never happens (like safety check error, or an error that occurs only when we made some mistake in writing code), I feel we should not panick.

File: pkg/trie/triedb/codec/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/codec/node.go]: The idiomatic go way would be to accept interface and return concrete types.

File: pkg/trie/triedb/codec/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/codec/node.go:125]: put a comment explaining this type?

File: pkg/trie/triedb/codec/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/codec/node.go:138]: panic is an attack vector in blockchain systems. Do we absolutely need it here or we can return an error?

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go:56]: can do without panic?

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go:61]: panic?

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go:67]: panic?

File: pkg/trie/triedb/triedb.go
Comment (kishansagathiya) [inline pkg/trie/triedb/triedb.go:833]: panic?

**Review by kishansagathiya — COMMENTED** [review]
- I don't like the use of panics. It's an attack vector in blockchains, we should avoid it as much as we can. I have tried to point out where I saw, but if there are more instances, please change it unless it's absolutely necessary.
- There are very few comments. We could do with a lot more comments here. I remember we needed a comment on everything exported. Is that not the case anymore? Did we remove that from `make lint`?

**Comment by dimartiro** [comment]
> * I don't like the use of panics. It's an attack vector in blockchains, we should avoid it as much as we can. I have tried to point out where I saw, but if there are more instances, please change it unless it's absolutely necessary.
> * There are very few comments. We could do with a lot more comments here. I remember we needed a comment on everything exported. Is that not the case anymore? Did we remove that from `make lint`?

Thanks Kishan, regarding the panics, I followed the same logic Parity has [here](https://github.com/paritytech/trie)
However since this code is not used yet, I want to revisit all this panics in a different PR, we already have an issue for that #3942, since it could imply some refactor and I don't want to block Tim now. 
So I gonna resolve all those comments and address them later in #3942

Regarding the godoc, I'll add the missing docs and check the linter to see why it isn't complying about it.

Thank you!

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go]: Comment. I can guess what this would be for, but I am not sure. I would nice to have comments here.

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go:232]: comment

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go]: What do you mean by reference in this context? Having comments for each exported things would be helpful here.

File: pkg/trie/triedb/node.go
Comment (kishansagathiya) [inline pkg/trie/triedb/node.go]:
```suggestion
	// Write children
```

File: pkg/trie/triedb/triedb.go
Comment (kishansagathiya) [inline pkg/trie/triedb/triedb.go:708]: comment

**Review by kishansagathiya — APPROVED** [review]
> I followed the same logic Parity has [here](https://github.com/paritytech/trie)

I see. They have used panics at places. Just 11 places though. Doesn't seem to be used that much. Anyways, not using them is just much safer.

> Regarding the godoc, I'll add the missing docs and check the linter to see why it isn't complying about it.

I am saying the opposite. Linter is complying despite having very few comments. I have now pointed out instances where it would make sense to have comments.

> However since this code is not used yet, I want to revisit all this panics in a different PR, we already have an issue for that https://github.com/ChainSafe/gossamer/issues/3942, since it could imply some refactor and I don't want to block Tim now.

This way we would be accumulating panics until #3942 is done. Ideally, we should have avoided panics in this PR. It would be best to prioritize #3942 right after this. 


Approving so that I don't end up blocking you, but add suggested comments before merging.

File: pkg/trie/triedb/triedb.go
Comment (dimartiro) [reply pkg/trie/triedb/triedb.go:708]: This is not exported, but I can add it anyways

**Pull request merged**
